[Merged by Bors] - p2p: use backoff during routing discovery

ivan4th opened the pull request:
## Motivation

Previous attempts at enabling routing discovery caused network overload on some of the nodes

## Description

Avoid excess DHT queries and connection attempts by using the same backoff mechanisms as those used by Kubo and peer discovery code in `go-libp2p-pubsub`

## Test Plan

Tested on a mainnet node

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update [changelog](../CHANGELOG.md) as needed

ivan4th committed:
p2p: use backoff during routing discovery

Avoid excess DHT queries and connection attempts

ivan4th committed:
Update CHANGELOG.md

fasmat commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -173,13 +194,27 @@ func New(h DiscoveryHost, opts ...Opt) (*Discovery, error) {
 	if len(d.bootnodes) == 0 {
 		d.logger.Warn("no bootnodes in the config")
 	}
+	rngSrc := rand.NewSource(rand.Int63())
+	cacheSize := 100
+	var err error
+	d.connBackoff, err = backoff.NewBackoffConnector(h, cacheSize, d.dialTimeout,
+		backoff.NewExponentialBackoff(d.minConnBackoff, d.maxConnBackoff, backoff.FullJitter,
+			time.Second, 5.0, 0, rand.New(rngSrc)))
```
That seems a bit convoluted, how about just passing a small wrapper around randv2?

```
type rngSource struct{}

func (r rngSource) Int63() int64 {
	return randv2.Int64()
}

func (r rngSource) Seed(seed int64) {}
```

```suggestion
	cacheSize := 100
	var err error
	d.connBackoff, err = backoff.NewBackoffConnector(h, cacheSize, d.dialTimeout,
		backoff.NewExponentialBackoff(d.minConnBackoff, d.maxConnBackoff, backoff.FullJitter,
			time.Second, 5.0, 0, rngSource{}))
```

ivan4th replied:
Fixed

fasmat commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -244,15 +286,28 @@ func (d *Discovery) Start() error {
+		var err error
 		d.disc = p2pdiscr.NewRoutingDiscovery(d.dht)
+		if d.discBackoff {
+			rng := rand.New(rand.NewSource(rand.Int63()))
+			d.disc, err = backoff.NewBackoffDiscovery(
+				d.disc,
+				backoff.NewExponentialBackoff(d.minBackoff, d.maxBackoff, backoff.FullJitter, time.Second, 5.0, 0, rng),
+			)
```
Same here:

```suggestion
			d.disc, err = backoff.NewBackoffDiscovery(d.disc,
				backoff.NewExponentialBackoff(
					d.minBackoff,
					d.maxBackoff,
					backoff.FullJitter,
					time.Second,
					5.0,
					0,
					rngSource{},
				),
			)
```

ivan4th replied:
Fixed

fasmat commented on p2p/dhtdiscovery/discovery_test.go:
```diff
@@ -62,7 +57,7 @@ func TestSanity(t *testing.T) {
 	require.NoError(t, err)
 	bootdisc.Start()
-	defer bootdisc.Stop()
+	t.Cleanup(bootdisc.Stop)
```
In my opinion `defer` is correct here. Its primarily a conceptual difference `defers` are executed just before the test ends (before `Test_XXX` returns), while `t.Cleanup` are executed after the test completes.

The question here is when should `bootdisc` be stopped? As part of the test or as part of the cleanup?

ivan4th replied:
Well on one hand `bootdisc.Stop()` belongs to the test
But re `t.Cleanup()`, the test will still fail if it fails, but it makes refactoring the test later much easier in most cases.
If you extract the piece of code into a "setup" function, it is very easy to shoot yourself in the foot with `defer`s, b/c they'll stop the services etc. immediately before the setup function returns. With `t.Cleanup`, this doesn't happen.
Also, `t.Cleanup()` does the right thing if subtests are added to the test later and these subtests use `t.Parallel()`

ivan4th replied:
(the latter thing about `t.Parallel()` is probably not exactly applicable to this particular case but still worth noting)

fasmat replied:
In a setup method I'd expect `bootdisc.Stop` to be called in a cleanup method that is returned by the setup method and deferred.

It depends on the specific case, but for parallel tests it's also imo safer to use `defer`:

```go
func TestMultiple(t *testing.T) {
	t.Run("test1", func (t *testing.T) {
		t.Parallel()

		data, cleanup := someSetup(t)
		defer cleanup()
		// Actual test
	})
	t.Run("test2", func (t *testing.T) {
		t.Parallel()

		data, cleanup := someSetup(t)
		defer cleanup()
		// Actual test
	})
}
```

if those `cleanup` calls are invoked with `t.Cleanup` it could happen by accident that they are using the `testing.T` of the main test which is less likely with `defer`.

Another reason to use `defer` over `t.Cleanup` is to avoid issues on Windows: if `t.Cleanup(bootdisc.Stop())` closes file handles to temporary files (created inside a `t.TempDir()`) it works on macOS and linux but on windows it will fail with an error that go cannot delete the temporary files because they have open file handles.

ivan4th replied:
Ok, I can change it, although it was more than once that I got bitten by `defer` during test refactoring and spent some time debugging the problem afterwards, as it is not always very obvious until you spot that `defer`

In parallel tests, the setup outside the subtests might use `t.Cleanup()` while setting up common slow-to-start fixture for the subtests. In this case the cleanup code must use "main" `testing.T`, and the cleanups will only get executed after all of the subtests finish:

```
func TestMultiple(t *testing.T) {
	data := someSetup(t) // uses t.Cleanup() for cleanup code
	t.Run("test1", func (t *testing.T) {
		t.Parallel()
		// Actual test
	})
	t.Run("test2", func (t *testing.T) {
		t.Parallel()
		// Actual test
	})
	// any cleanup done in `setup` gets executed only after
	// both `test1` and `test2` finish running
}
```

Re the Windows issue, `t.TempDir()` uses `t.Cleanup()` internally to delete the directory. Cleanups are executed strictly in reverse order, just like `defer`, so the only possibility I see is scheduling the cleanup func that closes the file _before_ invoking `t.TempDir()` somehow, which is a mistake by itself

ivan4th replied:
More re parallel: https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

ivan4th replied:
Changed to `defer` for now

fasmat replied:
> More re parallel: https://brandur.org/fragments/go-prefer-t-cleanup-with-parallel-subtests

To me this is kinda expected? `defer` is called before the test ends (so sub tests might still be running) while `t.Cleanup` is called after.

> Re the Windows issue, `t.TempDir()` uses `t.Cleanup()` internally to delete the directory. Cleanups are executed strictly in reverse order, just like `defer`, so the only possibility I see is scheduling the cleanup func that closes the file before invoking `t.TempDir()` somehow, which is a mistake by itself

That was not the issue. The issue was that the file wasn't closed when the cleanup happened, and to make sure that the close happened before the delete was to `defer` the call rather than to pass it to `t.Cleanup`. I do remember some tests where we had to change `t.Cleanup` to `defer` or it wouldn't pass on windows because of that very reason. I wasn't able to quickly come up with a simple example. Maybe this issue is also not present any more in Go 1.22

ivan4th committed:
Merge branch 'develop' into feature/slower-disc

ivan4th committed:
p2p: simplify rand usage in dhtdiscovery

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -444,100 +487,116 @@ func (d *Discovery) advertiseNS(ctx context.Context, ns string, active func() bo
+		select {
+		case <-ctx.Done():
+			return false, nil
+		case <-tickerCh:
+			if !active() {
+				return true, nil
+			}
```
It took me a while to understand that `active` is never `nil` here. Please, consider adding a comment explaining it.
```suggestion
		case <-tickerCh:
			// Note: ticker is not started if active is nil.
			if !active() {
				return true, nil
			}
```

ivan4th replied:
Makes sense, thanks

poszu commented on p2p/dhtdiscovery/discovery.go:
```diff
@@ -444,100 +487,116 @@ func (d *Discovery) advertiseNS(ctx context.Context, ns string, active func() bo
+	checkInterval time.Duration,
+	out chan<- peer.AddrInfo,
+) (cont bool, err error) {
+	if ctx.Err() != nil {
+		return false, err
+	}
```
```suggestion
	if ctx.Err() != nil {
		return false, ctx.Err()
	}
```

ivan4th committed:
Update p2p/dhtdiscovery/discovery.go

Co-authored-by: Bartosz Różański <bartek.roza@gmail.com>

ivan4th committed:
Update p2p/dhtdiscovery/discovery.go

Co-authored-by: Bartosz Różański <bartek.roza@gmail.com>

ivan4th committed:
p2p: handle discovery startup errors properly (#5888)

There was no error checking for DHT discovery startup, yet it could fail e.g. during LevelDB corruption:
```
Error: init services: error starting discovery: open leveldb at /opt/go-spacemesh/go-spacemesh-0/p2p: leveldb/journal: block/chunk corrupted: chunk length overflows block (1270 bytes) [file=000247.log]
```
In this case, error was being swallowed and DHT not started, breaking acquisition of new peers.

Co-authored-by: Ivan Shvedunov <ivan4th@users.noreply.github.com>

ivan4th committed:
p2p: dhtdiscovery: fix lint/test issues

ivan4th committed:
Merge branch 'develop' into feature/slower-disc

ivan4th commented:
bors merge
